Add pointer concept (#1724)
exercism/go: concepts/pointers/introduction.md

> # About

Reviewer: ```suggestion
# Introduction
```

# About

> Like many other languages, Go has pointers. A pointer hold the memory address of a value.

Reviewer 2: I think we need to give people a reason for why they're learning this before they work through such a complex topic. I've copied and shortened the bit at the bottom of document here. 

```suggestion
Like many other languages, Go has pointers. 
If you're new to pointers, they can feel a little mysterious but once you get used to them, they're quite straight-forward.
They're a crucial part of Go, so take some time to really understand them.

Before digging into the details, it's worth understanding the use of pointers. Pointers are a way to share memory with other parts of our program, which is useful for two major reasons:
1. When we have large amounts of data, making copies to pass between functions is very inefficient. 
  By passing the memory location of where the data is stored instead, we can dramatically reduce the resource-footprint of our programs.
2. By passing pointers between functions, we can access and modify the computer's memory directly, meaning that any changes made by one function are immediately visible to other parts of the program when the function ends.
```

## Variables and Memory

Let's say we have a regular integer variable `a`:

```go
var a int
```

> When we declare a variable, Go has to find a place in memory to place its value. Conviniently, when we need to fetch the value stored in that piece of memory, we can just refer to it by the variable name.

Reviewer 2: ```suggestion
When we declare a variable, Go has to find a place in memory to store its value. This is largely abstracted from us — when we need to fetch the value stored in that piece of memory, we can just refer to it by the variable name.
```

> For instance, when we do:
> 
> ```go
> a + 2
> ```
> 
> We are effectively fetching the value stored in the memory associated with the variable `a` and adding 2 to it.

Reviewer 2: ```suggestion
For instance, when we write `a + 2`, we are effectively fetching the value stored in the memory associated with the variable `a` and adding 2 to it.
```

Reviewer 2: I don't think this one needs to be split over multiple lines.

Similarly, when we need to change the value in the piece of memory of `a`, we can use the variable name to do an assignment:

```go
a = 3
```

> The piece of memory that `a` is associated with now will have the value `3`.

Reviewer 2: ```suggestion
The piece of memory that is associated with `a` will now will now be storing the value `3`.
```

## Pointers

> While variables allow us to refer to values in memory, sometimes it's useful to know the memory address of the value that the variable refers. Pointers hold the memory addresses of those values. A pointer declaration looks like this:

Reviewer 2: ```suggestion
While variables allow us to refer to values in memory, sometimes it's useful to know the **memory address** to which the variable is pointing. **Pointers** hold the memory addresses of those values. A pointer declaration looks like this:
```

I've just tweaked this so that we show why they're called pointers (I always find if you can make the naming seem "natural" when teaching something, it sinks in better)

```go
var p *int // 'p' can hold the memory address of an integer
```

> Here we declaring a variable `p` of type `*int`. The type `*int` means "pointer to int". In other words, we are saying that `p` is a pointer to int, which means that `p` will hold the memory address of an integer. The zero value of pointers is `nil`. A `nil` pointer holds no memory address.

Reviewer 2: ```suggestion
Here we declaring a variable `p` of type `*int`. The type `*int` means "pointer to int". In other words, we are saying that `p` is a pointer to int, which means that `p` will hold the memory address of an integer. The zero value of pointers is `nil` because a `nil` pointer holds no memory address.
```

Reviewer 2: I felt those two sentences should be linked so it feels a little less like two seperate facts.

### Getting a pointer to a variable

> The operator `&` gives the memory address of its operand. This means that to get the memory address of the value of a variable `a`, we can do `&a`. `&a` will create a pointer to the value of `a`:

Reviewer 2: I found this a little confusing, mainly because I'm not used to the word "operand". 

```suggestion
To find the memory address of the value of a variable, we can use the `&` operator. 
For example, if we want to find and store the member address of variable `a` in the pointer `p`, we can do the following:
```

Author: I was largely inspired by [Tour of Go: Pointers](https://tour.golang.org/moretypes/1) that also calls it operand.

But I agree the suggestion makes it a little more clear!

```go
var a int
a = 2

var p *int

> p = &a // the variable 'p' wil hold the memory address of 'a'

Reviewer 2: ```suggestion
p = &a // the variable 'p' will hold the memory address of 'a'
```

```

### Accessing the value via a pointer (dereferencing)

> When we have a pointer, we might want to know the value that is in the memory address of the pointer. We can do this using the `*` operator:

Reviewer 2: ```suggestion
When we have a pointer, we might want to know the value store in the memory address to which it points. We can do this using the `*` operator:
```

```go
var a int 
a = 2

var p *int 

> p = &a // the variable 'p' wil hold the memory address of 'a'

Reviewer 2: ```suggestion
p = &a // the variable 'p' will hold the memory address of 'a'
```

var b int
b = *p // b == 2
```

The operation `*p` fetches the value stored at the memory address stored in `p`. This operation is often called "dereferencing".

We can also use the derefering operator to assign a new value to the memory address referenced by the pointer:

```go
var a int
a = 2   // declare int variable 'a' and assign it the value of 2

var pa *int
pa = &a          // 'pa' now refers to the memory address of 'a'  
*pa = *pa + 2    // increment by 2 the value at memory address 'pa'

fmt.Println(a)   // Output: 4
                 // 'a' will have the new value that was changed via the pointer!
```

Assigning to `*pa` will change the value stored at the memory address `pa` holds. Since `pa` holds the memory address of `a`, by assigning to `*pa` we are effectively changing the value of `a`! 

> A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program panic at runtime!

Reviewer 2: ```suggestion
A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program panic at runtime!
```

Could we use a different word to "panic" here? Maybe "crash". I'm not sure someone coming from a different language (e.g. Ruby) would be familiar with the word, and so I have images in my mind of someone running around panicking, rather than the more technical usage :)

Author: In Go, a crash (in the sense of unhandled exception) is called a panic, that was the main reason for the use of panic there.

But I it doesn't hurt to make the introduction more inclusive and use a more general term.

Thanks for the suggestion.

Author: Fixed.

### Pointers to structs

So far we've only seen pointers to primitive values. We can also create pointers for structs:

```go
type Person struct {
    Name string
    Age  int
}

var peter Person
peter = Person{Name: "Peter", Age: 22}

var p *Person
p = &peter
```

> We could've also created a new `Person` and immediately get a pointer to it:

Reviewer 2: ```suggestion
We could have also created a new `Person` and immediately stored a pointer to it:
```

```go
func addPeterAge(m map[string]int){
	m["Peter"] = 22
}
```

If we create a map and call this function, the changes the function made to the map persist after the function ended. This is a similar behavior we get if we were using a pointer, but note how on this example we are not using any referencing/dereferencing or any of the pointer syntax:

```go
ages := make(map[string]int)
addPeterAge(ages)
fmt.Println(ages)
// Output: map[Peter:22]
// The changes the function 'addPeterAge' made to the map are visible after the function ends!
```

## Why are pointers useful?

Pointers are a way to share memory with other parts of our program. When you pass a pointer as an argument to a function, you are sharing the memory address of a value with that function. The function can then use that memory address to access and modify the value. All the modifications that the function makes to that value through the pointer are then visible to other parts of the program when the function ends. Thus, pointers can be a way to share memory.

> Another situation where sharing memory can be useful is to avoid copies of large amounts of data. If we have a piece of data that takes a lot of space in memory, it might not be practical to copy that data to every place we need it. In that situation, we might consider using a pointer and pass the pointer around instead. That way we avoid copies of large amounts of data, but we allow several parts of the program to access it via the pointer. 

Reviewer 2: We can probably remove this from here now, as we have a summary version at the top. However, it'd be good to have this slightly more detailed version stay in the about.md

Author: I like this change a lot!

I talked a bit with some people to know their troubles learning pointers before doing this exercise. One of the things they told me is that they had trouble learning pointers because they were not seeing why they were useful. That's why I made this little section, but I agree 100% that having this upfront is even better.

Change made.
